lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs

```diff
@@ -247,184 +233,110 @@ instance SoftDerivation ShelleyKey where
+        ShelleyKey $ xPrvChangePass oldPwd newPwd prv
+
+    publicKey = \(ShelleyKey prv) ->
+        ShelleyKey (toXPub prv)
+
+    digest = \(ShelleyKey pub) ->
+        hash (unXPub pub)
+
+    getRawKey =
+        getKey
+
+    keyTypeDescriptor _ = "she"
 
 {-------------------------------------------------------------------------------
                          Relationship Key / Address
 -------------------------------------------------------------------------------}
 
 instance PaymentAddress 'Mainnet ShelleyKey where
-    paymentAddress (ShelleyKey k0) =
-        encodeShelleyAddress (addrSingle @'Mainnet) [xpubPublicKey k0]
-    liftPaymentAddress (KeyFingerprint k0) =
-        encodeShelleyAddress (addrSingle @'Mainnet) [k0]
+    paymentAddress paymentK = do
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (enterprise + networkId)
+            putByteString . blake2b224 $ paymentK
+      where
+        enterprise = 96
+        networkId = 1
+
+    liftPaymentAddress (KeyFingerprint fingerprint) =
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (enterprise + networkId)
+            putByteString fingerprint
+      where
+        enterprise = 96
+        networkId = 1
 
 instance PaymentAddress ('Testnet pm) ShelleyKey where
-    paymentAddress (ShelleyKey k0) =
-        encodeShelleyAddress (addrSingle @('Testnet _)) [xpubPublicKey k0]
-    liftPaymentAddress (KeyFingerprint k0) =
-        encodeShelleyAddress (addrSingle @('Testnet _)) [k0]
+    paymentAddress paymentK =
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (enterprise + networkId)
+            putByteString . blake2b224 $ paymentK
+      where
+        enterprise = 96
+        networkId = 0
+
+    liftPaymentAddress (KeyFingerprint fingerprint) =
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (enterprise + networkId)
+            putByteString fingerprint
+      where
+        enterprise = 96
+        networkId = 0
 
 instance DelegationAddress 'Mainnet ShelleyKey where
-    delegationAddress (ShelleyKey k0) (ShelleyKey k1) =
-        encodeShelleyAddress (addrGrouped @'Mainnet) (xpubPublicKey <$> [k0, k1])
-    liftDelegationAddress (KeyFingerprint k0) (ShelleyKey k1) =
-        encodeShelleyAddress (addrGrouped @'Mainnet) ([k0, xpubPublicKey k1])
+    delegationAddress paymentK stakingK =
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (base + networkId)
+            putByteString . blake2b224 $ paymentK
+            putByteString . blake2b224 $ stakingK
+      where
+        base = 0
+        networkId = 1
+
+    liftDelegationAddress (KeyFingerprint fingerprint) stakingK =
+        Address $ BL.toStrict $ runPut $ do
+            putWord8 (base + networkId)
+            putByteString fingerprint
+            putByteString . blake2b224 $ stakingK
+      where
+        base = 0
+        networkId = 1
```

Comment:
network id =1 is only one possible mainnet, it can be form 1 to 15 (if I understand it well). we are ok now with fixed one value?

Reply:
I believe so. We have some time to see the other mainnets coming, I hope :no_mouth: